core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionValidator.scala

```diff
@@ -12,42 +13,49 @@ object ExpressionValidator {
   // presence: inputs which are both empty and optional are pre-validated, so they are skipped here
   // absence: validate all inputs normally
 
-  private[expressions] def validateAndParse(methodConfiguration: MethodConfiguration, inputsToParse: Map[String, AttributeString], allowRootEntity: Boolean, parser: SlickExpressionParser): ValidatedMethodConfiguration = {
+  private[expressions] def validateAndParse(methodConfiguration: MethodConfiguration, gatherInputsResult: GatherInputsResult, allowRootEntity: Boolean, parser: SlickExpressionParser): ValidatedMethodConfiguration = {
+    val inputsToParse = gatherInputsResult.processableInputs map { mi => (mi.workflowInput.localName.value, AttributeString(mi.expression)) }
     val (emptyOutputs, outputsToParse) = methodConfiguration.outputs.partition { case (_, expr) => expr.value.isEmpty }
 
-    val parsed = ExpressionParser.parseMCExpressions(inputsToParse, outputsToParse, allowRootEntity, parser)
+    val parsed = ExpressionParser.parseMCExpressions(inputsToParse.toMap, outputsToParse, allowRootEntity, parser)
 
     // empty output expressions are also valid
-    val validatedOutputs = emptyOutputs.keys.toSeq ++ parsed.validOutputs
+    val validatedOutputs = emptyOutputs.keys.toSet ++ parsed.validOutputs
 
-    ValidatedMethodConfiguration(methodConfiguration, parsed.validInputs, parsed.invalidInputs, validatedOutputs, parsed.invalidOutputs)
+    ValidatedMethodConfiguration(methodConfiguration, parsed.validInputs, parsed.invalidInputs, gatherInputsResult.missingInputs, gatherInputsResult.extraInputs, validatedOutputs, parsed.invalidOutputs)
   }
 
-  // validate a MC without retrieving the Method from the Repo: assume all inputs are required
-  def validateAndParseMCExpressions(methodConfiguration: MethodConfiguration, allowRootEntity: Boolean, parser: SlickExpressionParser): ValidatedMethodConfiguration =
-    validateAndParse(methodConfiguration, methodConfiguration.inputs, allowRootEntity, parser)
+  // validate a MC, skipping optional empty inputs, and return a ValidatedMethodConfiguration
+  def validateAndParseMCExpressions(methodConfiguration: MethodConfiguration,
+                                    gatherInputsResult: GatherInputsResult,
+                                    allowRootEntity: Boolean,
+                                    parser: SlickExpressionParser): ValidatedMethodConfiguration = {
+
+    val validated = validateAndParse(methodConfiguration, gatherInputsResult, allowRootEntity, parser)
+
+    // a MethodInput which is both optional and empty is already valid
+    val emptyOptionalInputs = gatherInputsResult.emptyOptionalInputs map { _.workflowInput.localName.value }
+    validated.copy(validInputs = validated.validInputs ++ emptyOptionalInputs)
```

Comment:
Maybe instead of doing a `copy`, we could add the `emptyOptionalInputs` up in `validateAndParse(...)`?

Reply:
I think there's _something_ fishy here. Given that `vapMCEs()` just calls `vap()` the difference between the two boils down to whether you want to be explicitly told "here are some inputs you failed to specify, but it turned out okay" or whether you just don't care to hear that information. Add to that the fact that `vap()` is _only_ called by `vapMCEs()` and I start to wonder why the distinction at all.

Thanks. I'll look into this.